logback-core/src/main/java/ch/qos/logback/core/BasicStatusManager.java

```diff
@@ -57,10 +59,15 @@ public void add(Status newStatus) {
         // LBCORE-72: fire event before the count check
         fireStatusAddEvent(newStatus);
 
-        count++;
-        if (newStatus.getLevel() > level) {
-            level = newStatus.getLevel();
-        }
+        count.increment();
+        int newLevel = newStatus.getLevel();
+        int currentLevel;
+        do {
+            currentLevel = level.get();
+            if (newLevel <= currentLevel) {
+                break;
+            }
+        } while (!level.compareAndSet(currentLevel, newLevel));
```

Comment:
How about
```java
synchronized (statusListLock) {
      count++;
       if (newStatus.getLevel() > level) {
           level = newStatus.getLevel();
       }
       if (statusList.size() < MAX_HEADER_COUNT) {
          statusList.add(newStatus);
      } else {
          tailBuffer.add(newStatus);
      }
  }
```

The code might be more simple.

Reply:
Yes, we certainly could take advantage of the existing lock that's going to be taken out.

However, for correctness, the reads would also need to use the same lock:

```java
    public int getLevel() {
        synchronized (statusListLock) {
            return level;
        }
    }

    public int getCount() {
        synchronized (statusListLock) {
            return count;
        }
    }
```
Otherwise the fields would need to be made volatile:

```java
    volatile int count = 0;
    volatile int level = Status.INFO;
```

It also occurred to me that perhaps the level should be reset when the count is reset:

```java
    public void clear() {
        synchronized (statusListLock) {
            count = 0;
            level = Level.INFO;
            statusList.clear();
            tailBuffer.clear();
        }
    }
```